keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,560 @@
+expect that a ClusterQueue's `borrowingLimit` for `credits` would be set to `0`
+to enforce the intended fiscal constraint.
+
+## Design Details
+
+Since Kueue's workload.Info struct is already a well-defined
+place where resource filtering is being applied to compute the effective resource
+request from a Workload's PodSpec, the core of the implementation is
+straightforward and follows from how `excludedResourcePrefixes` are already processed.
+
+The non-trivial design issues are how to configure the transformations,
+the possible granularity of that configuration, and how to maximize observability
+of the effective resources requested by a Workload.
+
+### Specifying the Transformation
+
+A transformation matrix is defined as a single object.  Most likely
+this object would be embedded inside Kueue's `Configuration` API, but
+it would also be possible to make it a separate top-level config API or 
+use a config map containing a single yaml field that would be deserialized
+into the type defined below.  To support finer-grained configuration, instead
+of defining a single controller-wide transformation matrix, they could be embedded
+in ClusterQueue ResourceGroups (perhaps implicitly by instead embedding the
+transformation matrix in a ResourceFlavor).
+
+```go
+type ResourceTransformationStrategy string
+const Retain ResourceTransformationStrategy = "Retain"
+const Replace ResourceTransformationStrategy = "Replace"
+
+type ResourceTransformation struct {
+  // key is the output resource name and value is the quantity of the output resource 
+  // that corresponds to 1 unit of the input resource
+  ConversionBaseValues map[corev1.ResourceName]resource.Quantity `json:"conversionBaseValues"`
+
+  // should the input resource be replaced or retained
+  Strategy ResourceTransformationStrategy `json:"strategy"`
+}
+
+// The map key is the input resource name
+type ResourceTransformationMatrix map[corev1.ResourceName]ResourceTransformation
+
+```
+By using maps instead of arrays, we avoid needing
+to define the semantics of multiple entries having the same input or output
+resource name.  Using the input resource name as the key to the outer map
+avoids conflicting definitions of `Replace` for an input resource.
+
+As an example, the following list of 4-tuples
+
+* (`nvidia.com/mig-1g.5gb`, `example.com/accelerator-memory`, `5Gi`, `replace`)
+* (`nvidia.com/mig-2g.10gb`, `example.com/accelerator-memory`, `10Gi`, `replace`)
+* (`nvidia.com/mig-4g.20gb`, `example.com/accelerator-memory`, `20Gi`, `replace`)
+* (`nvidia.com/mig-1g.5gb`, `example.com/credits`, `10`, `replace`)
+* (`nvidia.com/mig-2g.10gb`, `example.com/credits`, `15`, `replace`)
+* (`nvidia.com/mig-4g.20gb`, `example.com/credits`, `30`, `replace`)
+* (`cpu`, `example.com/credits`, `1`, `retain`)
+
+would be represented in yaml when embedded in the Kueue Configuration object as
+```yaml
+resources:
+  excludedResourcePrefixes: []
```

Comment:
What if the transformation resource name prefixs match this `excludedResourcePrefixes`?
The transformation rules are just ignored? Or will we introduce any validation rules?

Reply:
As implemented, `excludedResourcePrefixes` is applied first.  We probably should add a validation rule to flag this however.   Leaving this unresolved for now to remind me.

Reply:
My suggestion is to leave semantic validation of the `Resources` struct as separate item if we think it is needed.

There seem to be two main areas that could be checked:
1.  A resource that shows up as an `input` in `mappings` that would match a string in `excludedResourcePrefixes`.  This seems like it could be flagged as configuration error.  Probably giving `excludedResourcePrefixes` priority is the least surprising semantics.
2. Validation at the ClusterQueue level that nothing in a `coveredResources` list has a `Replace` resource mapping and that nothing in a `coveredResources` list matches an excludedResourcePrefixes.  Probably marking the ClusterQueue inactive and reporting the resason in the conditions message is the least surprising semantics.